server-common/src/main/java/org/apache/kafka/server/util/Deadline.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.util;
+
+import org.apache.kafka.common.utils.Time;
+
+import java.math.BigInteger;
+import java.util.Date;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+
+public class Deadline {
+    private final long nanoseconds;
+
+    public static Deadline fromMonotonicNanoseconds(
+        long nanoseconds
+    ) {
+        return new Deadline(nanoseconds);
+    }
+
+    public static Deadline fromDelay(
+        Time time,
+        long delay,
+        TimeUnit timeUnit
+    ) {
+        return fromDelay(time.nanoseconds(), delay, timeUnit);
+    }
+
+    public static Deadline fromDelay(
```

Comment:
Can this be private? Or if it needs to be accessible, maybe we can make it `fromDelayNs`